flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/cursor/stop/MessageIdStopCursor.java

```diff
@@ -43,8 +45,12 @@ public MessageIdStopCursor(MessageId messageId, boolean inclusive) {
         checkArgument(!earliest.equals(messageId), "MessageId.earliest is not supported.");
         checkArgument(!latest.equals(messageId), "Use LatestMessageStopCursor instead.");
 
-        this.messageId = convertToMessageIdImpl(messageId);
         this.inclusive = inclusive;
+        try {
+            this.messageId = MessageId.fromByteArray(messageId.toByteArray());
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
```

Comment:
Why not just convert it to a `MessageIdAdv`? The `fromByteArray` and `toByteArray` conversion actually loses some info for `BatchMessageIdImpl` (see the `acker` field).

```java
    public static MessageIdImpl convertToMessageIdImpl(MessageId messageId) {
        if (messageId instanceof BatchMessageIdImpl) {
            return (BatchMessageIdImpl) messageId;
        } else if (messageId instanceof MessageIdImpl) {
            return (MessageIdImpl) messageId;
        } else if (messageId instanceof TopicMessageIdImpl) {
            return convertToMessageIdImpl(((TopicMessageIdImpl) messageId).getInnerMessageId());
        }
    }
```

You can see before https://github.com/apache/pulsar/pull/19158, the implementation is above. In that PR, I uses `fromByteArray` and `toByteArray` just for `TopicMessageIdImpl`. After https://github.com/apache/pulsar/pull/19414, the recommended way is just to perform the conversion by casting to `MessageIdAdv`.

Reply:
We don't need that acker field indeed. BTW, this is just used for checking this is a valid message id. Flink will distribute the application to the downstream TMs which requires the serialization on messge id. The acker will be omitted finally.

Reply:
If you only need a `MessageIdImpl`, you can just use `MessageIdImpl` here to make it clear that the MessageId you want represents exactly a non-batched message.

But `MessageIdAdv` is a more stable interface that was introduced by a formal PIP, while `MessageIdImpl` is an implementation class. Using a class instead of an interface in ecosystem might make the upstream side hard to do some refactoring.

If you want to perform a validation that a message ID of a batch message is not allowed here, just check `msgIdAdv.getBatchIndex()`.

Reply:
I see. So I think we may just forget this complex details and use the `MessageId` directly. Let's hide the dark side of the Pulsar.

```suggestion
        this.messageId = messageId;
```

Reply:
Sounds reasonable :D